kit/src/components/message/mod.rs

```diff
@@ -368,11 +369,76 @@ fn ChatText(cx: Scope<ChatMessageProps>) -> Element {
 }
 
 pub fn markdown(text: &str) -> String {
-    // TODO: This does not work with code or pre elements.
-    //let txt = text.trim().replace(' ', "&nbsp;"); // need to do this else leading whitespaces are ignored
-    let parser = pulldown_cmark::Parser::new(text);
-    // Write to a new String buffer.
+    let txt = text.trim();
+
+    let mut options = Options::empty();
+    options.insert(Options::ENABLE_STRIKETHROUGH);
+
+    let mut text_lines: Vec<&str> = txt.split('\n').collect();
     let mut html_output = String::new();
-    pulldown_cmark::html::push_html(&mut html_output, parser);
-    html_output.replace("<p>", "").replace("</p>", "")
+    let mut in_paragraph = false;
+    let mut in_code_block = false;
+    let mut add_text_language = true;
+
+    for line in &mut text_lines {
```

Comment:
You shouldnt need to have a mutable reference here. You could just own `line` and use `mut line`. This would also remove the `mut text_lines` above when splitting.